### PR TITLE
Move crds to crds dir for helm3 and installCRDs flag for supporting h…

### LINKED ISSUE #286: Move crds to crds dir for helm3
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
- Move crds from `templates` to `crds` which is the recommended way for installing crds using helm3
- Add `template/crds.yaml` with 

  ```yaml
  {{- if .Values.installCRDs }}
  {{- range $path, $bytes := .Files.Glob "crds/*.yaml" }}
    {{ $.Files.Get $path }}
  ---
  {{- end }}
  {{- end }}
  ```
`installCRDs` by default will be false so crd install using helm2 is skipped. If using helm2 then `--set installCRDs=true`

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Kubernetes version: (use `kubectl version`):

/good-first-issue

### PR DESCRIPTION
**Pull request by Evalle**
**PR Summary**
- Moves crds to crds dir for helm3 and installCRDs flag for supporting helm2
- Fixes #286

**Special notes for the reviewer**:
@aramase  have moved crds to crds dir in `charts/secrets-store-csi-driver` should I do the same ing `manifest-staging` ?

### CODE REVIEW

**Comment by aramase** [comment]
@Evalle Thank you for the PR. All the changes are to be made in `manifest_staging/charts/secrets-store-csi-driver` only. That contains the latest charts and yamls that will be moved to `charts` at the time of release so we can ensure all chart and yaml changes are tagged with a release. So could you update the PR to make the changes in `manifest_changing` only?

Also, like I mentioned in the issue, we would need to add `template/crds.yaml` to make the charts work with helm2 as well. That blob would need to be under a conditional `installCRDs`, which users can set true for helm2 crd install. helm2 isn't aware of `crds` dir, so it would skip crd installation otherwise.

**Comment by Evalle** [comment]
@aramase thanks for the feedback. I've addressed your comment, could you please re-review? Thanks!

File: manifest_staging/charts/secrets-store-csi-driver/template/crds.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{{- if .Values.installCRDs }}
+{{- range $path, $bytes := .Files.Glob "crds/*.yaml" }}
+  {{ $.Files.Get $path }}
+---
+{{- end }}
+{{- end }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/template/crds.yaml:6]: nit: new line

**Comment by Evalle** [comment]
@aramase everything is done except for 

> and a note in the readme on --set installCRDs=true if using helm2 to install the chart.

In which `README.md`I should add it? There are three of them, two of which has a prerequisite `Helm 3.0`

Thanks

**Comment by aramase** [comment]
> In which README.mdI should add it? There are three of them, two of which has a prerequisite Helm 3.0

@Evalle It would be in `manifest_staging/charts/secrets-store-csi-driver/README.md`. We need instructions for helm2 and how `installCRDs=true` needs to be set only for helm2.

**Comment by Evalle** [comment]
@aramase  I've addressed all your comments, could you please re-review? Thx!

**Comment by ritazh** [comment]
@Evalle can you pls also update https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/Makefile#L162 to cp the files to the new `crds` folder instead?

**Comment by ritazh** [comment]
Can we also add an e2e test for helm 2 install and delete? I tested this locally with helm 2 install works great but delete does not remove the crds.

**Comment by aramase** [comment]
> Can we also add an e2e test for helm 2 install and delete? I tested this locally with helm 2 install works great but delete does not remove the crds.

Helm 2 doesn't remove the CRDs. We could add a note to say if the crds need to be deleted, then run `kubectl delete`?

**Comment by ritazh** [comment]
> We could add a note to say if the crds need to be deleted

Sure we can start with docs

**Comment by Evalle** [comment]
> can you pls also update https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/Makefile#L162 to cp the files to the new crds folder instead?

Done, thanks!

**Comment by aramase** [comment]
/hold

Thank you for the PR. The PR looks great!

One thing I tried was to do an upgrade of the chart with some changes in the CRD schema. Unfortunately, helm doesn't upgrade the CRDs which means the schema defined is still old. This could potentially break new driver versions that require the CRD schema to be updated. Ref: [Helm CRD lifecycle](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#:~:text=There%20is%20no%20support%20at,handle%20CRDs%20and%20their%20lifecycle.)

**Comment by aramase** [comment]
Some of the options we could potentially consider for supporting helm2 and helm3 charts -
1. Document installation of the CRD using `kubectl apply -f` after helm install/upgrade to ensure the latest CRD version is always being used. 
2. The helm recommendation is to use sub charts for CRDs.
3. Continue keeping the CRDs in `template/` dir. This works for helm2 and helm3, however this also removed the CRDs spec when the chart is uninstalled. This could mean loss of Custom resources on chart delete. 

I'm trying to play with some options to see best how we can include this change, but also open to other opinions.

**Comment by aramase** [comment]
@Evalle Could you add a note in the helm README.md in manifest_staging dir for installing the CRDs with 

```bash
kubectl apply -f deploy/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml
kubectl apply -f deploy/secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml
```

after installing the chart as helm3 currently doesn't manage lifecycle of CRDs?

**Comment by Evalle** [comment]
@aramase wow that was a lot of rebasing, but I think I've finished. Please re-review. Thx.

**Comment by aramase** [comment]
@Evalle Looks like the rebase isn't correct - https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/289/files#diff-53f51bf797df947afaf95576bd80491eR5. Could you rebase it right and update the PR?

**Comment by Evalle** [comment]
I see, let me do it again.

**Comment by Evalle** [comment]
@aramase Looks like I had an issue with VSCode, my bad. Could you please re-review?

File: manifest_staging/charts/secrets-store-csi-driver/README.md [hunk 2]
```diff
@@ -6,13 +6,22 @@ Quick start instructions for the setup and configuration of secrets-store-csi-dr
...
 ### Prerequisites
 
-- [Helm v3.0+](https://helm.sh/docs/intro/quickstart/#install-helm)
+- [Helm](https://helm.sh/docs/intro/quickstart/#install-helm)
 
 ### Installing the chart
 
+Helm v2.x
+```bash
+$ helm repo add secrets-store-csi-driver https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/master/charts
```
[the remark this answers is not in the document]
Comment (aramase) [reply manifest_staging/charts/secrets-store-csi-driver/README.md:15]: Could we update to this - 

### Installing the chart

Add the charts repo
```bash
$ helm repo add secrets-store-csi-driver https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/master/charts
```

Helm v2.x
```bash
$ helm install csi-secrets-store secrets-store-csi-driver/secrets-store-csi-driver --set installCRDs=true
```

Helm v3.0+
```bash
$ helm install csi-secrets-store secrets-store-csi-driver/secrets-store-csi-driver
```

Install CRDs after installing the chart
```bash
$ kubectl apply -f deploy/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml
$ kubectl apply -f deploy/secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml
```

File: manifest_staging/charts/secrets-store-csi-driver/template/crds.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{{- if .Values.installCRDs }}
+{{- range $path, $bytes := .Files.Glob "crds/*.yaml" }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/template/crds.yaml:2]: This file needs to be in `templates` dir

**Review by aramase — COMMENTED** [review]
Could you also remove this conditional - https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/manifest_staging/charts/secrets-store-csi-driver/templates/csidriver.yaml#L8. This will not work with helm2. The minimum chart version is `v1.16`, so that should be taken care of already.

**Comment by Evalle** [comment]
@aramase could you please re-re-re-re-review :) ?

**Comment by aramase** [comment]
@Evalle The most recent changes look good. I'll add this to the agenda for tomorrow's [community call](https://docs.google.com/document/d/1q74nboAg0GSPcom3kLWCIoWg43Qg3mr306KNL58f2hg/edit?usp=sharing) to see if other folks have any comments before merging the PR.

**Comment by aramase** [comment]
As discussed in the community call, `v0.0.15` has updates to CRDs and since helm3 doesn't handle upgrades of CRDs we are planning to move this to release `v0.0.16` which will not contain any CRD changes.

**Review by aramase — COMMENTED** [review]
@Evalle Are you interested in picking up this PR again? We're planning to include the move to helm3 in the next release `v0.1.0`. The helm charts will only be supported with helm3, so we no longer need the `installCRDs` flag and `crds.yaml` in the templates folder. 

Essentially the required changes are: 
1. Move the 2 CRDs to crds directory
2. Update the `apiVersion` from v1 to v2 in [here](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/manifest_staging/charts/secrets-store-csi-driver/Chart.yaml#L1).

Let us know if you would like to rebase and update this PR.

**Comment by Evalle** [comment]
Progress:
- [x] remove installCRDs flag
- [x]  remove crds.yaml in the `templates` folder.
- [ ] Move the 2 CRDs to `crds` directory - @aramase should I create the `crds` folder in the root? CRDs are in  `secrets-store-csi-driver/manifest_staging/charts/secrets-store-csi-driver/crds/` now
- [x] Update the apiVersion from v1 to v2

**Comment by aramase** [comment]
>  Move the 2 CRDs to crds directory - @aramase should I create the crds folder in the root? CRDs are in secrets-store-csi-driver/manifest_staging/charts/secrets-store-csi-driver/crds/ now

Looks like you've the `crds` directory setup correctly in the root.

**Review by aramase — COMMENTED** [review]
/lgtm

**Comment by ritazh** [comment]
Thank you for completing this PR!

Has there been more discussion/thoughts around this concern? Perhaps a pre-upgrade hook? IMO, we should figure this out before cutting v0.1.0.

> Helm will install the crds in the crds/ directory if the crds are not already installed. If the crds are installed, it will not validate that the version installed matches the served/stored versions specified in the manifests. On upgrade it doesn't touch the crds/ directory at all.

**Pull request merged**